### ISSUE TITLE
Metadata Merge I: what is the combined metadata for column names

### ISSUE THREAD

**Issue by iherman** [w3c/csvw#76]
(My apologies if this is a bit long; I will have two more issues that will refer back to this example):

Consider the following case. The data is (I shortened one of our usual examples):

```
GID,On Street,Species
1,ADDISON AV,Celtis australis
2,EMERSON ST,Liquidambar styraciflua
```

Per our specification, this yields a conceptual (embedded) metadata of the form:

```
{
  "schema" : {
    "columns" : [{
      "name" : "GID"
    },{
      "name" : "On Street"
    },{
      "name" : "Species"
    }]
  }
}
```

Let us also suppose that we have the metadata in a file of the form (I again shorten our example):

```
{
  "schema" : {
    "columns" : [{
      "name" : "GID"
    },{
      "title" : "On Street",
      "name"  : "on-street"
    },{
      "title" : "Species",
      "name" : "species"
    }]
}
```
## what is the combined metadata?

The model document specifies that the metadata should be merged in this order:
1. conceptual metadata
2. file metadata

But it is not clear what merging means. In the example above, does it mean that the result is something like:

```
{
  "schema" : {
    "columns" : [{
      "name" : "GID"
    },{
      "title" : "On Street",
      "name"  : "On street"
    },{
      "title" : "Species",
      "name"  : "Species"
    }]
}
```

Is this what we want? Or is it so that the conceptual (embedded) metadata should only define the "title" properties and not the "name" properties? 

Note that this is also related to issue #33 about the default column name issue. Indeed, if there is _no_ column name given, the system would generate some artificial one "Col 1", "Col 2", etc. However, if we follow the rules, those names will _stick_ because they have a higher priority than any other setting. I guess that is definitely _not_ what we want:-(

**Comment by JeniT** [comment]
You are mixing up different levels of processing here. No conceptual schema gets generated from the embedded metadata. Rather the CSV file is parsed into a table model that has a table T with:
- three columns:
  1. with the `title` `GID` containing cells [1,1] and [2,1]
  2. with the `title` `On Street` containing cells [1,2] and [2,2]
  3. with the `title` `Species` containing cells [1,3] and [2,3]
- two rows:
  1. containing cells [1,1], [1,2] and [1,3]
  2. containing cells [2,1], [2,2] and [2,3]
- six cells:
  - [1,1] with the value `1`
  - [1,2] with the value `ADDISON AV`
  - [1,3] with the value `Celis australis`
  - [2,1] with the value `2`
  - [2,2] with the value `EMERSON ST`
  - [2,3] with the value `Liquidambar styraciflua`

Then, when the processor interprets the metadata, this model is augmented with additional annotations based on the information in the metadata document. Afterwards, it looks like:
- three columns:
  1. with the `title` `GID` and the `name` `GID` containing cells [1,1] and [2,1]
  2. with the `title` `On Street` and the `name` `on-street` containing cells [1,2] and [2,2]
  3. with the `title` `Species` and the `name` `species` containing cells [1,3] and [2,3]
- two rows:
  1. containing cells [1,1], [1,2] and [1,3]
  2. containing cells [2,1], [2,2] and [2,3]
- six cells:
  - [1,1] with the value `1`
  - [1,2] with the value `ADDISON AV`
  - [1,3] with the value `Celis australis`
  - [2,1] with the value `2`
  - [2,2] with the value `EMERSON ST`
  - [2,3] with the value `Liquidambar styraciflua`

I assume that this isn't clear from the documents; would including an example like this help?

I'm also not sure that this is answering your fundamental point which is presumably "what happens when the embedded column title isn't one of those listed in the metadata". It's an error, but I think we probably want to recover nicely from that error by retaining the original title (as well as that from the metadata) in the annotated model. So if, with the same schema, the CSV looked like:

```
GID,Street,Species
1,ADDISON AV,Celtis australis
2,EMERSON ST,Liquidambar styraciflua
```

with the second column called `Street` instead of `On Street` then the second column in the model would have a `title` annotation with an array of values (it's possible to have multiple titles), `["Street", "On Street"]`.

(I think I'd expect the RDF generated from this to include a `rdfs:label` triple for whatever property was used for `on-street` for each of the titles that was provided.)

**Comment by iherman** [comment]
I think there are several issues here that gets intertwined...

The reason I got to this and to what I called 'conceptual' metadata and, primarily, what to do if there are no proper names given to columns. Actually, that is handled in issue #33 (which we have almost closed), so that part of the issue may be moot. (Sorry if I did not bind this properly to issue #33).

But the more general question for me is how one combines the various metadata, which is really issue #77. For a better management of the issues, I would respond on that thread then... see you there:-)

**Comment by 6a6d74** [comment]
@JeniT: the explicit example really helps clarify. Thanks.

**Comment by gkellogg** [comment]
Still confused:

The model doc says to process in order: 
1. metadata supplied by the user
2. metadata embedded within the tabular data file
3. metadata in a document linked to using a Link header
4. file-specific metadata
5. directory-specific metadata

As I understand it, (2) is determined by extracting headers from the CSV (if any) and setting them as Title metadata. When processing, I open the CSV. Based on that, I either get Metadata from Link header, from file-specific, or directory-specific. After that, I merge in metadata from the CSV itself (right now, just column titles) and then merge in any user-specified metadata.

An ambiguity (or processing-order problem) here is that if the user-supplied metadata says there are no header rows (or a different number), then I need to re-process the CSV to get new header values, or undo the previous process of extracting titles from the CSV.

It would be simpler to say that if there is user-supplied metadata, it is the _only_ metadata used. Otherwise, it is extracted using steps 2-5 above.

**Comment by gkellogg** [comment]
Resolved in PR #169 and on http://www.w3.org/2015/01/28-csvw-minutes.html.

**Comment by gkellogg** [comment]
RESOLUTION: We will resolve #43, #76, #77, #78 as suggested in https://lists.w3.org/Archives/Public/public-csv-wg/2015Jan/0050.html
